File: C Programs ☠️/detectCycle_SLL.c
// Write a C program to detect a loop in a single linked list.

#include "singlylinkedlist.h"

void detectCycle(NODE *);

int main()
{
    NODE *head = NULL;
    printf("Please Enter More Than 3 data\n");
    head = createList(head);
    displayList(head);
    detectCycle(head);

    NODE *tail = head;
    int size = 1;
    while (tail->next != NULL)
    {
        tail = tail->next;
        size++;
    }

    if (size >= 3)
    {
        tail->next = head->next->next;
        printf("\nLinked List Successfully Created with Cycle");
        detectCycle(head);
    }
    return 0;
}

void detectCycle(NODE *head)
{
    if (head == NULL || head->next == NULL)
    {
        printf("\nNo Cycle in Linked List");
        return;
    }

    NODE *slow = head;
    NODE *fast = head;
    NODE *entry = head;

    while (fast->next && fast->next->next)
    {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast)
        {
            while (slow != entry)
            {
                slow = slow->next;
                entry = entry->next;
            }
            printf("\nCycle Detected at Node Data: %d", entry->data);
            return;
        }
    }
    printf("\nNo Cycle in Linked List");
}

// OUTPUT
// Please Enter More Than 3 data
// Creating Linked List...
// Enter -1 to end
// Enter the data: 16
// Enter the data: 51
// Enter the data: 29
// Enter the data: 47
// Enter the data: -1
// HEAD -> 16 -> 51 -> 29 -> 47 -> NULL

// No Cycle in Linked List
// Linked List Successfully Created with Cycle
// Cycle Detected at Node Data: 29

File: Python Programs 🐍/detectCycle_SLL.py
# Write a python program to detect
# a loop in a single linked list.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def createList(self):
        print("Creating Linked List...")
        print("Enter -1 to end")
        num = int(input("Enter the data: "))
        while(num != -1):
            newNode = Node(num)
            if self.head is None:
                self.head = newNode
            else:
                last = self.head
                while last.next:
                    last = last.next
                last.next = newNode
            num = int(input("Enter the data: "))

    def displayList(self):
        if self.head is None:
            print("List is empty")
            return
        print("HEAD", end='')
        temp = self.head
        while temp:
            print(" -> " + str(temp.data), end='')
            temp = temp.next
        print(" -> NULL\n")

    def detectCycle(self):
        slow = self.head
        fast = self.head
        entry = self.head
        if fast is None and fast.next is None:
            print("No Cycle in Linked List")
            return
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                while slow != entry:
                    slow = slow.next
                    entry = entry.next
                print("Cycle Detected at Node Data: "+str(entry.data))
                return
        print("No Cycle in Linked List")


if __name__ == "__main__":
    myList = LinkedList()
    print("Enter More Than 2 Data")
    myList.createList()
    myList.displayList()
    myList.detectCycle()

    size = 1
    tail = myList.head
    while tail.next:
        tail = tail.next
        size += 1

    if size >= 2:
        tail.next = myList.head.next
        print("Linked List Successfully created with Cycle")
        myList.detectCycle()


# OUTPUT
# Enter More Than 2 Data
# Creating Linked List...
# Enter - 1 to end
# Enter the data: 51
# Enter the data: 97
# Enter the data: 3
# Enter the data: 6
# Enter the data: -1
# HEAD -> 51 -> 97 -> 3 -> 6 -> NULL

# No Cycle in Linked List
# Linked List Successfully created with Cycle
# Cycle Detected at Node Data: 97
